scipy/stats/stats.py

```diff
@@ -4153,6 +4154,179 @@ def binary_search(n, n1, n2, side):
     return oddsratio, pvalue
 
 
+def _barnard_exact_compute_gammaln_combination(n):
+    """Compute all log combination of C(n, k)"""
+    gammaln_arr = special.gammaln(np.arange(n + 1) + 1)
+    return special.gammaln(n + 1) - gammaln_arr - gammaln_arr[::-1]
+
+
+def barnard_exact(
+        table,
+        alternative='two-sided',
+        pooled=True,
+        num_it=4):
+    """
+    Perform a Barnard exact test on a 2x2 contingency table.
+
+    Parameters
+    ----------
+    table : array_like of ints
+        A 2x2 contingency table.  Elements should be non-negative integers.
+    alternative : {'two-sided', 'less', 'greater'}, optional
+        Defines the alternative hypothesis.
+        The following options are available (default is 'two-sided'):
+
+          * 'two-sided'
+          * 'less': one-sided
+          * 'greater': one-sided
+
+    pooled : bool, optional
+        Whether to compute score statistic with pooled variance or
+        unpooled (Wald statistic)
+
+    num_it : int, optional
+        Number of iterations of the grid search. Default is 4. Must be
+        non-negative.
+
+    Returns
+    -------
+    score_statistic : float
+        Z statistic with pooled (Score) or unpooled (Wald) variance, depending
+        on the user choice of `pooled` param.
+    p_value : float
+        P-value, the probability of obtaining a distribution at least as
+        extreme as the one that was actually observed, assuming that the
+        null hypothesis is true.
+
+    See Also
+    --------
+    chi2_contingency : Chi-square test of independence of variables in a
+        contingency table.
+    fisher_exact : Fisher exact test on a 2x2 contingency table.
+
+    References
+    ----------
+    .. [1] `Mehta, Cyrus & Senchaudhuri, Pralay. (2003). Conditional
+    versus Unconditional Exact Tests for Comparing Two Binomials`__.
+
+    .. __: https://www.statsols.com/hubfs/Resources_/Comparing-Two-Binomials.pdf
+
+    Examples
+    --------
+    examples from [1]_.
+
+    Consider the following example of a vaccine efficacy study (Chan, 1998).
+    In a randomized clinical trial of 30 subjects, 15 were innoculated with a
+    recombinant DNA influenza vaccine and the 15 were innoculated with a
+    placebo. Twelve of the 15 subjects in the placebo group (80%) eventually
+    became infected with influenza whereas for the vaccine group, only 7 of
+    the 15 subjects (47%) became infected. The data are tabulated as a 2 × 2
+    table::
+
+                    Vaccine  Placebo
+        Yes             8        2
+        No              1        5
+
+    We use this table to find the p-value:
+    >>> import scipy.stats as stats
+    >>> TX, pvalue = stats.barnard_exact([[7, 12], [8, 3]])
```

Comment:
Please return a dataclass object rather than a tuple. Tuples are difficult to extend in a backwards compatible way, so we are moving away from them.

Reply:
I am not sure to understand correctly. Do you want my function to return a named tuples ?

Reply:
[Dataclasses](https://docs.python.org/3/library/dataclasses.html) are a close cousin to named tuple which are more recent (3.7+ only); instead of accessing the result by index; you would return an object with a `TX` and `pvalue` field.

The reason being if you return a name tuple it is hard to add fields as people may use destructuring:

```
t,p_value = barnard_exact(....)
```

While with a dataclass you "force" them to use attribute access:

```
stat = barnard_exact(...)
t, p_value = stat.TX, stats.pvalue
```

Now you are free to add any other fields if it make sens.

the `@dataclass` is a class decorator that will automatically implement __init__, __repr__, and many other methods base on class level type annotation. 

The documentation is pretty straightforward; and if you feel uncomfortable or unsure let us know and we can help you do it.

Reply:
No, i was suggesting a [`dataclass`](https://docs.python.org/3/library/dataclasses.html). Any very simple class is OK, but creating a dataclass takes just one line.

Reply:
Going to change it then

Reply:
Resolved in 3cf1e6d917825c0321b43f5d8f473b198acbb390